crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_final_literal.rs

```diff
@@ -0,0 +1,128 @@
+use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, comparable::ComparableExpr};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+use crate::fix::snippet::SourceCodeSnippet;
+
+/// ## What it does
+/// Checks for redundant `Final[Literal[]]` annotations.
+///
+/// ## Why is this bad?
+/// A `Final[Literal[x]]` annotation can be replaced with just `Final`.
+///
+/// ## Example
+///
+/// ```python
+/// x: Final[Literal[42]]
+/// # or,
+/// x: Final[Literal[42]] = 42
+/// ```
+///
+/// Use instead:
+/// ```python
+/// x: Final = 42
+/// ```
+#[violation]
+pub struct RedundantFinalLiteral {
+    literal: SourceCodeSnippet,
+}
+
+impl Violation for RedundantFinalLiteral {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RedundantFinalLiteral { literal } = self;
+        if let Some(literal) = literal.full_display() {
+            format!("`Final[Literal[{literal}]]` can be replaced with a bare `Final`")
+        } else {
+            format!("`Final[Literal[...]] can be replaced with a bare `Final`")
+        }
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let RedundantFinalLiteral { literal } = self;
+        if let Some(literal) = literal.full_display() {
+            Some(format!(
+                "Replace `Final[Literal[{literal}]]` with a bare `Final`"
+            ))
+        } else {
+            Some(format!("Replace `Final[Literal[...]] with a bare `Final`"))
+        }
+    }
+}
+
+/// PYI064
+pub(crate) fn redundant_final_literal(checker: &mut Checker, ann_assign: &ast::StmtAnnAssign) {
+    let ast::StmtAnnAssign {
+        value: assign_value,
+        annotation,
+        ..
+    } = ann_assign;
+
+    let ast::Expr::Subscript(ast::ExprSubscript {
+        slice: literal_slice,
+        ..
+    }) = &**annotation
+    else {
+        return;
+    };
+    let ast::Expr::Subscript(ast::ExprSubscript { slice: literal, .. }) = &**literal_slice else {
+        return;
+    };
+
+    // Discards tuples like `Literal[1, 2, 3]`
+    // and complex literals like `Literal[{1, 2}]`
+    if !matches!(
```

Comment:
can potentially be moved to `ruff_python_ast/src/helpers.rs`, as a `is_literal` helper.